tests/harness/utils.py

```
@@ -74,6 +74,32 @@ def get_server_metrics_url(self, *server_types: str) -> list[str]:
+        """Return the inbound headers the mock server saw on inference paths.
+
+        Each entry is `{"path": "...", "headers": {<lowercased name>: <value>}}`.
+        Useful for asserting end-to-end header fidelity through the aiohttp
+        transport. Call `clear_recorded_headers()` first to scope the result
+        to a single test (the mock server is shared across the package).
+        """
+        import urllib.request
+
+        with urllib.request.urlopen(
+            f"{self.url}/test/recorded_headers", timeout=10
+        ) as resp:
```

Ruff (0.15.21)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)